From 9677fe4a6d6bd553fecffcfe4706926fd622c439 Mon Sep 17 00:00:00 2001
From: "qipeng.zha" <qipeng.zha@intel.com>
Date: Tue, 26 Aug 2014 23:16:38 +0800
Subject: [PATCH 2/2] PMIC: add "id" for intel_soc_pmic_set_pdata()

BZ:

Merge below patch from MCG-main,
Ic4d4ab4dce9802039b213281379a49af48b6c158

Sometime one platform driver can be used for several platfor
devices, thus pmic should support a way to add different platform
data to different platform devices which have the same name

Change-Id: I5887c9d6bc2ebd8c79ba00bd591bb61120117d83
Signed-off-by: qipeng.zha <qipeng.zha@intel.com>
---
 drivers/mfd/intel_soc_pmic_core.c  |    8 +++++---
 drivers/mfd/intel_soc_pmic_dc.c    |    6 +++---
 include/linux/mfd/intel_soc_pmic.h |    2 +-
 3 files changed, 9 insertions(+), 7 deletions(-)

diff --git a/drivers/mfd/intel_soc_pmic_core.c b/drivers/mfd/intel_soc_pmic_core.c
index a52c8ffd5ef9..86d8d8462dbf 100644
--- a/drivers/mfd/intel_soc_pmic_core.c
+++ b/drivers/mfd/intel_soc_pmic_core.c
@@ -35,6 +35,7 @@ struct cell_dev_pdata {
 	const char		*name;
 	void			*data;
 	int			len;
+	int			id;
 };
 static LIST_HEAD(pdata_list);
 
@@ -145,7 +146,7 @@ err:
 }
 EXPORT_SYMBOL(intel_soc_pmic_update);
 
-int intel_soc_pmic_set_pdata(const char *name, void *data, int len)
+int intel_soc_pmic_set_pdata(const char *name, void *data, int len, int id)
 {
 	struct cell_dev_pdata *pdata;
 
@@ -158,7 +159,7 @@ int intel_soc_pmic_set_pdata(const char *name, void *data, int len)
 	pdata->name = name;
 	pdata->data = data;
 	pdata->len = len;
-
+	pdata->id = id;
 	list_add_tail(&pdata->list, &pdata_list);
 
 	return 0;
@@ -467,7 +468,8 @@ int intel_pmic_add(struct intel_soc_pmic *chip)
 
 	for (i = 0; pmic->cell_dev[i].name != NULL; i++) {
 		list_for_each_entry(pdata, &pdata_list, list) {
-			if (!strcmp(pdata->name, pmic->cell_dev[i].name)) {
+			if (!strcmp(pdata->name, pmic->cell_dev[i].name) &&
+					(pdata->id == pmic->cell_dev[i].id)) {
 				pmic->cell_dev[i].platform_data = pdata->data;
 				pmic->cell_dev[i].pdata_size = pdata->len;
 			}
diff --git a/drivers/mfd/intel_soc_pmic_dc.c b/drivers/mfd/intel_soc_pmic_dc.c
index 069b792db385..d32cd5495904 100644
--- a/drivers/mfd/intel_soc_pmic_dc.c
+++ b/drivers/mfd/intel_soc_pmic_dc.c
@@ -402,7 +402,7 @@ static void dc_xpwr_chrg_pdata(void)
 	platform_init_chrg_params(&pdata);
 
 	intel_soc_pmic_set_pdata("dollar_cove_charger",
-				(void *)&pdata, sizeof(pdata));
+				(void *)&pdata, sizeof(pdata), 0);
 }
 
 static int fg_bat_curve[] = {
@@ -446,7 +446,7 @@ static void dc_xpwr_fg_pdata(void)
 		pdata.bat_curve[i] = fg_bat_curve[i];
 
 	intel_soc_pmic_set_pdata("dollar_cove_battery",
-				(void *)&pdata, sizeof(pdata));
+				(void *)&pdata, sizeof(pdata), 0);
 }
 
 static void dc_xpwr_pwrsrc_pdata(void)
@@ -461,7 +461,7 @@ static void dc_xpwr_pwrsrc_pdata(void)
 	pdata.en_chrg_det = true;
 
 	intel_soc_pmic_set_pdata("dollar_cove_pwrsrc",
-				 (void *)&pdata, sizeof(pdata));
+				 (void *)&pdata, sizeof(pdata), 0);
 }
 
 static int dollar_cove_init(void)
diff --git a/include/linux/mfd/intel_soc_pmic.h b/include/linux/mfd/intel_soc_pmic.h
index a3614b5f7411..f6c7dbbe6a3b 100644
--- a/include/linux/mfd/intel_soc_pmic.h
+++ b/include/linux/mfd/intel_soc_pmic.h
@@ -60,7 +60,7 @@ int intel_soc_pmic_writeb(int reg, u8 val);
 int intel_soc_pmic_setb(int reg, u8 mask);
 int intel_soc_pmic_clearb(int reg, u8 mask);
 int intel_soc_pmic_update(int reg, u8 val, u8 mask);
-int intel_soc_pmic_set_pdata(const char *name, void *data, int len);
+int intel_soc_pmic_set_pdata(const char *name, void *data, int len, int id);
 struct device *intel_soc_pmic_dev(void);
 
 #endif	/* __INTEL_SOC_PMIC_H__ */
-- 
1.7.9.5

